### PR TITLE
Remove default non-negativity constraint on least core subsidy

### LINKED ISSUE #294: Remove unnecessary constraint on Least Core subsidy
The Least Core is defined as:

$$
\begin{array}{lll}
\min e \in \mathbb{R}\\
\text{s.t. } & \sum_{i \in N} x_i = u(N), &\\
& \sum_{i \in S} x_i + e \ge u(S), & \forall S \subset N, S \neq \emptyset
\end{array}
$$

The inequality constraint is only defined on proper, non-empty subsets of the set of all agents/players/data points.

However, in some papers, e.g. [If You Like Shapley Then You’ll Love the Core](http://procaccia.info/wp-content/uploads/2020/12/core.pdf), it is defined on all subsets i.e. $\forall S \subseteq N$ and that implies that the subsidy must be non-negative $e \geq 0$ and that should not be the case.

### PR DESCRIPTION
**Pull request by AnesBenmerzoug**
### Description

This PR closes #294 

### Changes

- Add `non_negative_subsidy` flag to least core methods  with a default value of False.
- Remove empty subset from lower bound constraints, if found, because it is the same as using the constraint $e >= 0$ if $u(\emptyset) = 0$ (which is true in almost all cases) (i.e. setting `non_negative_subsidy=True`)
- Remove total utility from lower bound constraints, if found, because it is the same as using the constraint $e >= 0$ (i.e. setting `non_negative_subsidy=True`)
- Rename `options` to `solver_options`.
- Pass `solver_options` as dictionary instead of as kwargs.
- Shows a deprecation warnings when passing solver options as kwargs.
- Change default solver from ECOS to SCS.
- Set default solver max iterations to 10000.
- Update least core notebook.
- Update core and least core docs' constraints to use proper non-empty subsets.

### Checklist

- [x] Wrote Unit tests (if necessary)
- [X] Updated Documentation (if necessary)
- [x] Updated Changelog
- [X] If notebooks were added/changed, added boilerplate cells are tagged with `"nbsphinx":"hidden"`

### CODE REVIEW

**Comment by mdbenito** [comment]
(without looking at the PR) Note that the constraint for the empty set only implies e > 0 if u(S)=0

**Comment by mdbenito** [comment]
And, since you update the docs, would it be possible to keep the notation consistent with the rest of the docs? i.e. identify points $x_i$ with their indices $i$ and use $v_i$ instead of $v_u(x_i)$

**Comment by AnesBenmerzoug** [comment]
@mdbenito I actually used the same notation as in the rest of the data valuation documentation.

Do you mean only in the docstring? If not, then I would suggest we leave that as a separate issue and discuss it tomorrow.

**Comment by mdbenito** [comment]
> @mdbenito I actually used the same notation as in the rest of the data valuation documentation.

My bad, it's the rest of the doc that's quite inconsistent.
 
> Do you mean only in the docstring? If not, then I would suggest we leave that as a separate issue and discuss it tomorrow.

See #308

**Review by mdbenito — COMMENTED** [review]
Just had a couple of questions. Otherwise this looks good. We should repeat the experiments, btw...

File: src/pydvl/value/least_core/common.py [hunk 1]
```diff
@@ -47,23 +52,56 @@ def lc_solve_problem(
...
+        # Because we already have the equality constraint.
+        mask = np.ones_like(b_lb, dtype=bool)
+        mask[total_utility_indices] = False
+        b_lb = b_lb[mask]
+        A_lb = A_lb[mask]
+
+    # Remove the row(s) corresponding to the empty subset
+    # because it is the same as using the constraint e >= 0
+    # (i.e. setting non_negative_subsidy = True)
+    emptyset_utility_indices = np.where(A_lb.sum(axis=1) == 0)[0]
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/value/least_core/common.py:91]: Fixed.

File: src/pydvl/value/least_core/common.py [hunk 2]
```diff
@@ -47,23 +52,56 @@ def lc_solve_problem(
...
+    if len(total_utility_indices) == 0:
         b_eq = np.array([u(u.data.indices)])
     else:
-        b_eq = b_lb[total_utility_index]
+        b_eq = b_lb[total_utility_indices]
+        # Remove the row(s) corresponding to the total utility
+        # from the lower bound constraints
+        # because it is the same as using the constraint e >= 0
+        # (i.e. setting non_negative_subsidy = True)
+        # Because we already have the equality constraint.
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/value/least_core/common.py:82]: Fixed.

File: src/pydvl/value/least_core/__init__.py [hunk 3]
```diff
@@ -66,14 +71,28 @@ def compute_least_core_values(
...
     .. versionadded:: 0.5.0
     """
     progress: bool = kwargs.pop("progress", False)
 
+    # TODO: remove this before releasing version 0.7.0
+    if kwargs:
+        warnings.warn(
+            "Passing solver options as kwargs was deprecated in 0.6.0 and will be removed in 0.7.0 "
+            "Use solver_options instead.",
+            DeprecatedWarning,
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/value/least_core/__init__.py]: Nice catch! I updated this.

File: src/pydvl/value/least_core/common.py [hunk 4]
```diff
@@ -109,7 +147,8 @@ def lc_solve_problems(
     algorithm: str,
     config: ParallelConfig = ParallelConfig(),
     n_jobs: int = 1,
-    **options,
+    non_negative_subsidy: bool = False,
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply src/pydvl/value/least_core/common.py:168]: That's true. What do you suggest in this case? Setting it to True for the next release and then set it back to False for the one after that?

[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/value/least_core/common.py:168]:
```suggestion
    non_negative_subsidy: bool = True,
```

**Review by mdbenito — APPROVED** [review]

**Pull request merged**